ci: SW-739 created shared workflows for building/testing/releasing dotnet-based projects

andrew-ubco opened the pull request:
- this uses secrets for `GH_USERNAME`, `GH_TOKEN` and `GH_ACTIONS_SHARED_SERVICES_ROLE_ARN` directly - is this ok?
- not sure about the names of all the workflow files, but we need to differentiate between these and others (eg. for SAM, frontend projects, etc)

An (untested) example of how to use the shared workflows is in this draft PR: https://github.com/UBCO-Bikes/plutus/pull/10

andrew-ubco committed:
ci: SW-739 created shared workflows for building/testing/releasing dotnet-based projects

Roland-UBCO commented on .github/workflows/_shared_multi-architecture-dotnet-build-and-test.yml:
```diff
@@ -0,0 +1,114 @@
+      image-name:
+        required: true
+        type: string
+      architecture:
```
Lets change this input type from `string` to `choice`, and then we can explicitly set the allowed options (`arm64`, `amd64`?) https://github.blog/changelog/2021-11-10-github-actions-input-types-for-manual-workflows/

andrew-ubco replied:
Yep, I didn't realise we had `choice` (but admittedly I hadn't looked)

andrew-ubco replied:
As mentioned in slack, `choice` isn't supported for reusable workflows

Roland-UBCO commented on .github/workflows/_shared_multi-architecture-dotnet-build-and-test.yml:
```diff
@@ -0,0 +1,114 @@
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
I know we do this in lots of other places (which we should tidy up), but lets start using explicit versions here -> `ubuntu-20.04`

Roland-UBCO commented on .github/workflows/onpush-dotnet-test-and-create-release.yml:
```diff
@@ -0,0 +1,31 @@
+    env:
+      GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
+    steps:
+      - uses: rymndhng/release-on-push-action@v0.25.0
```
Need to check this action's behaviour.. will it still work as expected when in a shared workflow, and not a 'on push' workflow? How does it know which commit to add the release tag to?? I guess it will just use the current commit for the workflow that call this one? Will have to test to find out.
The action documentation is a bit thin regarding this.. https://github.com/rymndhng/release-on-push-action

andrew-ubco replied:
Yep, will check

andrew-ubco replied:
I've tested this and it works

Roland-UBCO commented on .github/workflows/onpush-dotnet-test-and-create-release.yml:
```diff
@@ -0,0 +1,31 @@
+name: Unit test and create new release
```
Need to think about filenames here. This is no longer an `onpush` action, so feels strange to keep calling it this. Maybe just needs to be `dotnet_test-and-create-release-version.yml`?

If we are going to have multiple shared workflows in here for different project types, maybe the naming convention should include the technology stack as the prefix, e.g `dotnet_`?

andrew-ubco replied:
The rationale for leaving onpush/onrelease in the names is that's where it will generally be called from - and yes, I did include `dotnet` in the name 😄

andrew-ubco replied:
Files renamed

Roland-UBCO commented on .github/workflows/onrelease-dotnet-build-amd64.yml:
```diff
@@ -0,0 +1,22 @@
+name: Build amd64 Docker Image
```
This workflow is now exactly the same as `build-arm64`, with the exception of the hardcoded architecture value. Could we collapse these two into a single workflow, where the `architecture` is passed in as a param? The project workflows can then just call this twice, with a different param each time?

File could then be called `dotnet_build-and-push-docker-image`?

andrew-ubco replied:
Yep

andrew-ubco replied:
Done

Roland-UBCO commented on .github/workflows/pull-request-dotnet-unit-tests.yml:
```diff
@@ -0,0 +1,18 @@
+name: Pull Request Unit Tests
```
Proposed name: `dotnet_run-unit-tests`

andrew-ubco replied:
Done

andrew-ubco committed:
refactor: changes after draft PR review:

- file renaming
- build workflow consolidation
- fixed hosted runner ubuntu version

andrew-ubco committed:
refactor: fixed workflow names

andrew-ubco committed:
chore: reverted to using string for architecture input as choice isn't supported in reusable workflows

andrew-ubco committed:
refactor: renamed dotnet build and test job
